rep-2001.rst

```diff
@@ -38,26 +38,45 @@ We define three main entry points for ROS users.
  * desktop (recommended)
  * ros_base
  * ros_core
+ * ros_minimal
 
 Variants
 --------
 
+ROS Minimal
+'''''''''''
+
+The `ros_minimal` variant is a highly restricted set of ROS2 functionality.
+It targets resource- and tool-constrained environments where package managers may be limited or unavailable.
+The intention is that this variant can target cross-compiling for custom OS sysroots that may not have build tools or Python available that run on the target platform.
```

Comment:
If it is an explicit goal for this variant to be able to be used on a target system that doesn't have Python available, a different building + environment setup process should be highlighted as the currently recommended one (colcon build + source a setup file) does not fulfill this requirement (https://github.com/colcon/colcon-core/issues/73)

Reply:
The goal isn't to have this be _built on_ a target system that doesn't have Python available. The intention is to be built _for_ a target system that doesn't have Python available. I want to make a strong distinction between the build-time environment and the run-time environment. While these are often the same in classic ROS workflows, It shouldn't be the assumption. Do you think I should add language around that here? I fully expect this to be built with colcon and linted with cpplint, for example.

Reply:
The build also needs Python to run the code generators for messages / services / actions.

Reply:
But that's not needed at runtime, right?

Reply:
> The intention is to be built for a target system that doesn't have Python available. 

That's what I meant as well but it was not clear.
Currently AFAICT it is *not* possible to source a colcon setup file (top-level setup.bash file generated by colcon) without Python: https://github.com/colcon/colcon-core/issues/73.
So for such a variant to be usable on a system that doesn't have Python available, we need a different workflow that allow to setup the environment on the target even if the target doesn't have Python. It can be modifying colcon to generate Python-agnostic setup files, or a different mechanism.

Reply:
Oh - I see what you mean. The way I have used it is to
* use `--merged-install`
* `export LD_LIBRARY_PATH=install/lib`
* invoke binaries by full path

Now that I'm thinking about that, yes, it seems we'll need a non-python way to set up the proper environment variables. Would it be possible to get away with creating a fairly simple pure-shell utiility? Something that would
* Set `LD_LIBRARY_PATH`
* Add all package `lib` folders to `PATH`

Reply:
That simple approach will only work for a subset. A complete shell-based solution must replicate the same logic as it is implemented in Python. That includes collecting all packages in the prefix, getting their dependencies (available in the filesystem, doesn't need to parse the manifests), order them topologically, and then source the setup files of each package.

If the shell-based solution doesn't implement the logic around `.dsv` files that would also imply a much longer time to setup the environment.

Reply:
Are these details just necessary to handle the different types of installs, or is it deeper than that? Perhaps we could impose limitations on the usage of this variant so as to avoid more complexity in the environment setup.

>  source the setup files of each package.

Do these require python, or just the top level script does?

Reply:
> Are these details just necessary to handle the different types of installs, or is it deeper than that? Perhaps we could impose limitations on the usage of this variant so as to avoid more complexity in the environment setup.

That has nothing do to with the different types of installs.

The importance is the order. One package might rely on an environment change caused by another package it depends on.

> > source the setup files of each package.

> Do these require python, or just the top level script does?

No, they don't.